### PR TITLE
Fixes #6636, #6657, #6694 - add field length validations by db size, extend audit field length

### PR DESCRIPTION
**Pull request by orrabin**
Fixes #6657: Major/monor version fields should not accept negative integer values while creating OS

### CODE REVIEW

**Comment by orrabin** [comment]
[test] again

**Comment by ohadlevy** [comment]
please update the commit message to be: fixes ####, fixes #### otherwise redmine will pick up only the first one.

**Comment by ohadlevy** [comment]
Also, usually, even if it seems to be annoying, it could be better to split to separate commits. (e.g. the db migration is not relate directly to the validations).

**Comment by domcleal** [comment]
@ohadlevy I think Redmine's fine with "fixes ####, ####" IME

File: test/unit/operatingsystem_test.rb [hunk 1]
```diff
@@ -233,4 +233,52 @@ class OperatingsystemTest < ActiveSupport::TestCase
     assert_equal operatingsystems(:suse), operatingsystems.first
   end
 
+  test "should create os with a name of 255 characters" do
+    os = FactoryGirl.build(:operatingsystem, :name => 'a' * 255)
+    assert os.valid?
```
Comment (domcleal) [inline test/unit/operatingsystem_test.rb]: we have helpers now for assert_valid, refute_valid etc which could be useful to shorten these tests

File: test/unit/operatingsystem_test.rb [hunk 2]
```diff
@@ -233,4 +233,52 @@ class OperatingsystemTest < ActiveSupport::TestCase
...
+    assert os.save
```
Comment (domcleal) [inline test/unit/operatingsystem_test.rb:239]: Avoid saving unless we need to (testing the validation is enough), it'll slow down test runs

**Comment by orrabin** [comment]
@domcleal Thanks, I changed the tests.

**Comment by orrabin** [comment]
@ohadlevy Should I change the comment?

**Comment by orrabin** [comment]
I used the gem: https://github.com/rubiety/validates_lengths_from_database
this also fixes #6658, #4153, #6712

File: app/models/user.rb [hunk 3]
```diff
@@ -11,6 +11,7 @@ class User < ActiveRecord::Base
   ANONYMOUS_ADMIN = 'foreman_admin'
   ANONYMOUS_API_ADMIN = 'foreman_api_admin'
 
+  validates_lengths_from_database  :except => [:firstname, :lastname, :format, :mail, :login]
```
Comment (domcleal) [inline app/models/user.rb:14]: Why the mix of using validates_length.. and individual :length declarations?  Should they be removed?

File: Gemfile [hunk 4]
```diff
@@ -20,6 +20,8 @@ gem 'oauth'
 gem 'deep_cloneable', '~> 2.0'
 gem 'foreigner', '~> 1.4.2'
 
+gem 'validates_lengths_from_database'
```
Comment (domcleal) [inline Gemfile]: I suggest pinning either with '~> 0.2.0' or ~> 0.2, just in case.

**Comment by domcleal** [comment]
I closed off the duplicate issues, you can just do: "fixes #6636, #6657, #6694 - add field length validations, extend audit field length" or similar.

File: app/models/user.rb (hunk 3 above)
Comment (orrabin) [inline app/models/user.rb:14]: @domcleal they aren't limited in the database and I was trying to avoid a migration.

Comment (domcleal) [inline app/models/user.rb:14]: Ah yes, it's an artificial restriction, thanks.

**Comment by domcleal** [comment]
Merged as 3034e8e for Foreman 1.6, thanks @orrabin!

**Pull request closed without merging**